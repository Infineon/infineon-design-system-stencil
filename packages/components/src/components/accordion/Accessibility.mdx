import { Meta, Description } from '@storybook/blocks';
import * as AccordionStories from './accordion.stories';
import { data } from './accessibility.ts';
import { getCompliancePercentageOfComponent, getMaxRating } from '../../global/utils/accessibility.ts';

export const componentName = "Accordion";

export const compliancePercentage =  getCompliancePercentageOfComponent(data);
export const componentRating = getMaxRating(compliancePercentage);

export const statusIcons = {
  'Passing': '✅',
  'Failing': '❌',
  'Not Applicable': '⚪'
};

<Meta of={AccordionStories}/>

# Accessibility of {componentName}
Overall this component has a rating of **{componentRating}**.

## Compliance Percentage
The table below shows the percentage of accessibility rules that are fulfilled for each level (A, AA, AAA).

| Level | Percentage |
|-------|------------|
| A     | {(compliancePercentage.A*100).toFixed(0)}% |
| AA    | {(compliancePercentage.AA*100).toFixed(0)}% |
| AAA   | {(compliancePercentage.AAA*100).toFixed(0)}% |

## WCAG rules
<table>
  <thead>
    <tr>
      <th>#</th>
      <th>Guideline</th>
      <th>Level</th>
      <th>Summary</th>
      <th>Status</th>
      <th>Comment</th>
    </tr>
  </thead>
  <tbody>
    {data.map((rule, index) => (
      <tr key={index}>
        <td>{rule.guidelineNumber}</td>
        <td>{rule.guideline}</td>
        <td>{rule.level}</td>
        <td>{rule.summary}</td>
        <td>{statusIcons[rule.status] || rule.status}</td>
        <td>{rule.comment}</td>
      </tr>
    ))}
  </tbody>
</table>

## Notes on Usage
### Set AriaLevel
Set the `AriaLevel` attribute to the correct level to ensure that the screen reader reads the content in the correct order.  
It refers to the heading level of the caption.
 