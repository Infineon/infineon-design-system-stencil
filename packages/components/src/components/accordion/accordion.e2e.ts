import { newE2EPage } from '@stencil/core/testing';

describe('ifx-accordion', () => {
  let page;
  let element;

  beforeEach(async () => {
    page = await newE2EPage();
    await page.setContent('<ifx-accordion></ifx-accordion>');
    element = await page.find('ifx-accordion');
  });

  it('renders', async () => {
    expect(element).toHaveClass('hydrated');
  });

  it('autoCollapse prop works as expected', async () => {
    element.setProperty('autoCollapse', true);
    await page.waitForChanges();
    expect(await element.getProperty('autoCollapse')).toBe(true);
  });

});
