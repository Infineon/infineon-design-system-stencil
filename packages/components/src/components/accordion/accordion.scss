// ifxAccordion.scss
@use "~@infineon/design-system-tokens/dist/tokens";
@use "../../global/font.scss";

:host {
  display: block;
}

.accordion-wrapper {
  display: flex;
  flex-direction: column;
  gap: 8px;
}