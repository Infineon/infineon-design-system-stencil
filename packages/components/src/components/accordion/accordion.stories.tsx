import { action } from '@storybook/addon-actions';

export default {
  title: 'Components/Accordion',
  component: 'ifx-accordion',
  //tags: ['autodocs'],

  args: {
    autoCollapse: false,
  },

  argTypes: {
    amountOfItems: { control: 'number' },
  },
};

const Template = args => {
  const accordionElement = document.createElement('ifx-accordion');
  const initialItem = document.createElement('ifx-accordion-item');
  initialItem.setAttribute('caption', `Label`);
  initialItem.setAttribute('open', `true`);

  initialItem.innerHTML = `
  Content for Initial Item. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent volutpat, ligula eu aliquam bibendum, orci nisl cursus ipsum, nec egestas odio sapien eget neque.
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent volutpat, ligula eu aliquam bibendum, orci nisl cursus ipsum, nec egestas odio sapien eget neque.
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent volutpat, ligula eu aliquam bibendum, orci nisl cursus ipsum, nec egestas odio sapien eget neque.
`;
  initialItem.addEventListener('ifxItemOpen', action('ifxItemOpen'));
  initialItem.addEventListener('ifxItemClose', action('ifxItemClose'));

  accordionElement.append(initialItem);

  accordionElement.setAttribute('auto-collapse', args.autoCollapse);
  for (let i = 1; i < args.amountOfItems; i++) {
    const item = document.createElement('ifx-accordion-item');
    item.setAttribute('caption', `Label`);
    item.setAttribute('open', `false`);

    item.innerHTML = `
        Content for Item #${i + 1}. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent volutpat, ligula eu aliquam bibendum, orci nisl cursus ipsum, nec egestas odio sapien eget neque.
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent volutpat, ligula eu aliquam bibendum, orci nisl cursus ipsum, nec egestas odio sapien eget neque.
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent volutpat, ligula eu aliquam bibendum, orci nisl cursus ipsum, nec egestas odio sapien eget neque.
    `;
    item.addEventListener('ifxItemOpen', action('ifxItemOpen'));
    item.addEventListener('ifxItemClose', action('ifxItemClose'));

    accordionElement.append(item);
  }

  return accordionElement;
};

export const Default = Template.bind({});
Default.args = {
  amountOfItems: 3,
};
