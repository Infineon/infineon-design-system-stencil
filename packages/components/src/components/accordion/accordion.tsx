//ifxAccordion.tsx
import { Component, h, Listen, Element, Prop } from '@stencil/core';

@Component({
  tag: 'ifx-accordion',
  styleUrl: 'accordion.scss',
  shadow: true,
})
export class Accordion {
  @Element() el: HTMLElement;
  @Prop() autoCollapse: boolean = false;

  @Listen('ifxItemOpen')
  async onItemOpen(event: CustomEvent) {
    if(this.autoCollapse) { 
      const items = Array.from(this.el.querySelectorAll('ifx-accordion-item'));
      for (const item of items) {
        const itemElement = item as HTMLIfxAccordionItemElement;
        if (itemElement !== event.target && (await itemElement.isOpen())) {
          itemElement.close();
        }
      }
    }
  }

  render() {
    return (
      <div class="accordion-wrapper">
        <slot />
      </div>
    );
  }
}
