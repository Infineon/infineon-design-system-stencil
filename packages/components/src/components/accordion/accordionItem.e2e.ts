import { newE2EPage } from '@stencil/core/testing';

describe('ifx-accordion-item', () => {
  let page;
  let element;

  beforeEach(async () => {
    page = await newE2EPage();
    await page.setContent('<ifx-accordion-item caption="Test"></ifx-accordion-item>');

    element = await page.find('ifx-accordion-item');
  });

  it('renders', async () => {
    expect(element).toHaveClass('hydrated');
  });

  it('displays the caption', async () => {
    const captionElement = await page.find('ifx-accordion-item >>> .accordion-caption');
    expect(captionElement.innerText).toEqual('Test');
  });

  it('expands and collapses on click', async () => {
    const titleElement = await page.find('ifx-accordion-item >>> .accordion-title');

    // Initially closed
    expect(await element.getProperty('open')).toBeFalsy();

    await titleElement.click();

    // Open after first click
    expect(await element.getProperty('open')).toBeTruthy();

    await titleElement.click();

    // Closed after second click
    expect(await element.getProperty('open')).toBeFalsy();
  });
});
