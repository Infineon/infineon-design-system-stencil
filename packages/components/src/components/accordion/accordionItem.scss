// ifxAccordionItem.scss
@use "~@infineon/design-system-tokens/dist/tokens";
@use "../../global/font.scss";

// :host {
//   display: block;
// }

.accordion-item {
  border-radius: 3px;
  overflow: hidden;
  transition: all 0.3s;
}

.accordion-title:hover {
  border: 1px solid tokens.$color-gray-200;
  background-color: tokens.$color-gray-200;
}

.accordion-title {
  display: flex;
  align-items: center;
  padding: 1rem;
  color: tokens.$color-default-500;
  border: 1px solid tokens.$color-gray-200;
  cursor: pointer;
}

.accordion-caption {
  font-weight: tokens.$font-weight-semi-bold;
  font-size: tokens.$font-size-m;
}

.accordion-content {
  max-height: 0;
  overflow: hidden;
  transition: max-height 0.3s ease-in-out;
}

.inner-content {
  padding: 24px;
  white-space: pre-wrap;
  /* wraps text at spaces and within words */
  word-wrap: break-word;
  /* breaks text within a word if necessary */
  overflow-wrap: anywhere;
  /* breaks text at arbitrary points when needed */
}

.accordion-icon {
  font-weight: bold;
  display: inline-block;
  transition: transform 0.3s;
  margin-right: 0.5rem;
}

.accordion-item.open .accordion-icon {
  transform: rotate(-180deg);
}