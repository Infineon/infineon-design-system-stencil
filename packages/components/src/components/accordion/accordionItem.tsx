//ifxAccordionItem
import { Component, Prop, h, State, Event, EventEmitter, Watch } from '@stencil/core';

@Component({
  tag: 'ifx-accordion-item',
  styleUrl: 'accordionItem.scss',
  shadow: true,
})
export class IfxAccordionItem {
  @Prop() caption: string;
  @Prop({
    mutable: true,
  }) open: boolean = false;
  @State() internalOpen: boolean = false;
  @Event() ifxItemOpen: EventEmitter;
  @Event() ifxItemClose: EventEmitter;
  private contentEl!: HTMLElement;


  componentWillLoad() {
    this.internalOpen = this.open;
  }

  componentDidLoad() {
    this.openAccordionItem()
  }

  componentDidUpdate() {
    this.openAccordionItem()
  }



  @Watch('open')
  openChanged(newValue: boolean) {
    this.internalOpen = newValue;
  }

  toggleOpen() {
    this.internalOpen = !this.internalOpen;
    this.open = this.internalOpen;

    if (this.internalOpen) {
      this.ifxItemOpen.emit({ isOpen: this.internalOpen });
    } else {
      this.ifxItemClose.emit({ isClosed: !this.internalOpen });
    }
  }

  openAccordionItem() {
    if (this.internalOpen) {
      this.contentEl.style.maxHeight = `${this.contentEl.scrollHeight}px`;
    } else {
      this.contentEl.style.maxHeight = '0';
    }
  }

  handleSlotChange(e) {
    const slotElement = e.target;
    const nodes = slotElement.assignedNodes();
    
    if(nodes.length > 0) {
      nodes.forEach(node => {
        const observer = new MutationObserver((mutationsList, _) => {
          for(let mutation of mutationsList) {
            if (mutation.type === 'childList') {
              if (this.internalOpen) {
                this.openAccordionItem();
              }
            }
          }
        });
        observer.observe(node, { attributes: true, childList: true, subtree: true });
      });
    }

    if (this.internalOpen) {
      this.openAccordionItem();
    }
  }

  render() {
    return (
      <div aria-label={this.caption} class={`accordion-item ${this.internalOpen ? 'open' : ''}`}>
        <div class="accordion-title" onClick={() => this.toggleOpen()}>
          <span class="accordion-icon">
            <ifx-icon icon="chevron-down-12" />
          </span>
          <span class="accordion-caption">{this.caption}</span>
        </div>
        <div class="accordion-content" ref={(el) => (this.contentEl = el as HTMLElement)}>
          <div class="inner-content">
            <slot onSlotchange={(e) => this.handleSlotChange(e)} />
          </div>
        </div>
      </div>
    );
  }
}
