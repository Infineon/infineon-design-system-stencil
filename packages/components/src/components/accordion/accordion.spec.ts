// src/components/accordion/accordion.spec.ts
import { newSpecPage } from 'jest-stencil-runner';
import { Accordion } from './accordion';


describe('ifx-accordion', () => {
 it('renders with correct structure', async () => {
  const { root } = await newSpecPage({
    components: [Accordion],
    html: `<ifx-accordion></ifx-accordion>`,
  });

  expect(root).toMatchSnapshot();
});
   
});