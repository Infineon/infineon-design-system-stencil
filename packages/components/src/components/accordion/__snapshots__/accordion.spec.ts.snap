// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ifx-accordion renders with correct structure 1`] = `
<ifx-accordion>
  <template shadowrootmode="open" shadowrootdelegatesfocus>
    <div class="accordion-wrapper">
      <slot>
      </slot>
    </div>
  </template>
</ifx-accordion>
`;
